### PR TITLE
Explicit device malloc

### PR DESCRIPTION
**Pull request by abadams**
Made Buffer::copy_to_device do something not-broken in JIT-using code.

This previously failed in mysterious ways. Copying to device in JIT code before any pipeline had run on device didn't properly initialize device state, and crashed on null target machine.

Even after a pipeline had run on device there was no way to feed copy_to_device the same Target as used in a subsequent realize (which is necessary to set the Debug flag), and there was no way to get it to automatically pick a device api that matched realize's default - so you have a schedule with a generic gpu_tile, but you need an explicit halide_cuda_device_interface().

This is a start, and is better than trunk, but there are still major unresolved problems with JIT and GPU. For example if you say:

b.device_malloc(Target("host-cuda"));

Then b gets used in an OpenCL pipeline, the underlying device allocation was freed and replaced with an OpenCL one, but it was freed in a copy of b's buffer struct. So b now contains a dev field that's a dangling pointer. We get a crash in b's destructor.

The root cause of these issues is the private copy of the user's Buffer object that we make. I have a separate branch I'm working on that lets Halide pipelines safely carry around references to user buffers instead of making private copies of them like this. The implementation is not pretty, but I think it should be the least surprising thing for users.

### CODE REVIEW

**Comment by abadams** [comment]
The unresolved problems I referred to are addressed in #1694 

These can be merged in either order. This one is probably less contentious. 

The travis failure appears to be a flake.

File: src/runtime/HalideBuffer.h [hunk 1]
```diff
@@ -37,6 +37,10 @@ struct halide_dimension_t {
 
 namespace Halide {
 
+class Target;
+halide_device_interface_t *get_device_interface_for_target(const Target &);
```
Comment (zvookin) [inline src/runtime/HalideBuffer.h]: Don't the function decls need an "extern"?

File: src/DeviceInterface.cpp [hunk 2]
```diff
@@ -46,6 +47,32 @@ bool lookup_runtime_routine(const char *name, const Target &target,
 
 }
 
+namespace Halide {
+halide_device_interface_t *get_device_interface_for_target(const Target &t) {
```
Comment (zvookin) [inline src/DeviceInterface.cpp]: Either passing just a feature flag or using DeviceAPI instead of a Target seems safer here.

Comment (abadams) [inline src/DeviceInterface.cpp]: The comment claims that you have to pass it a target with exactly one set. If you set multiple it uses the same ordering preference as the gpu_tile, etc. Not sure if this is the best behavior, but it's important that:

    foo.copy_to_device();
    bar.gpu_tile(x, y, 8, 8);

does the right thing. Maybe it should take a DeviceAPI instead?

**Comment by zvookin** [comment]
I guess my complaint is this seems like a fairly low-level routine and I'd
like the concept of default device to only exist very close to the coding
model. You could also rename the routine "get_default_device_interface_for_
target".

-Z-

**Comment by abadams** [comment]
Updated with more nuanced API. Unfortunately a Target is still required given a DeviceAPI in order to construct the shared runtime if necessary - the original code effectively used "host-cuda" or similar, which dropped Debug flags.

File: src/DeviceInterface.cpp [hunk 3]
```diff
@@ -146,35 +198,20 @@ int halide_weak_device_free(void *user_context, struct buffer_t *buf) {
...
 
 const struct halide_device_interface_t *halide_cuda_device_interface() {
-    Target target(get_host_target());
-    target.set_feature(Target::CUDA);
-    struct halide_device_interface_t *(*fn)();
-    if (lookup_runtime_routine("halide_cuda_device_interface", target, fn)) {
-        return (*fn)();
-    }
-    return nullptr;
+    Target t = get_jit_target_from_environment().with_feature(Target::CUDA);
```
Comment (dsharletg) [inline src/DeviceInterface.cpp]: Should this be get_host_target, like it was before? It's a bit sketchy for this API to depend on an environment variable (which get_jit_target_from_environment does). Also, this comment would be obsoleted if we changed get_device_interface_for_device_api above.

File: src/DeviceInterface.cpp [hunk 4]
```diff
@@ -46,6 +48,56 @@ bool lookup_runtime_routine(const char *name, const Target &target,
 
 }
 
+namespace Halide {
+halide_device_interface_t *get_default_device_interface_for_target(const Target &t) {
+    return get_device_interface_for_device_api(DeviceAPI::Default_GPU, t);
+}
+
+halide_device_interface_t *get_device_interface_for_device_api(const DeviceAPI &d, const Target &t) {
```
Comment (dsharletg) [inline src/DeviceInterface.cpp]: I wonder if this function should perhaps not take a target, and instead used get_host_target().with_feature(Target::x) for lookup_runtime_routine, where x depends on the value of d.

**Comment by abadams** [comment]
In both of these cases, it has to match the target you intend to pass to
Func::realize. Otherwise it'll clobber your choices for Target::Debug.
Using get_host_target was a bug in that regard, and meant that Debug didn't
work if you called device_malloc before you run a pipeline.

Basically the things that take Targets here should have the same default
value as Func::realize.

**Comment by abadams** [comment]
(get_jit_target_from_environment is still sketchy in the second usage, but it's less sketchy than get_host_target).

**Comment by dsharletg** [comment]
Makes sense, and I now see you already thought about this, sorry for the repeat question.

**Pull request merged**